### PR TITLE
Change Arc to use compare_exchange instead of compare_and_swap

### PR DESCRIPTION
**Pull request by Amanieu**
This should slightly improve code generation on architectures with a weak memory model like ARM.

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+ 9a28d4edc9375e5bf606c453d1e03a45ae8be0af

Nice!

**Comment by alexcrichton** [comment]
Looks like the slaves disappeared for whatever mysterious reason they do so, but otherwise all tests passed, so merging manually.

File: src/liballoc/arc.rs [hunk 1]
```diff
@@ -256,11 +256,11 @@ impl<T: ?Sized> Arc<T> {
...
-            // This Relaxed is OK because we're checking the value in the CAS
-            // below.
-            let cur = this.inner().weak.load(Relaxed);
+        // This Relaxed is OK because we're checking the value in the CAS
+        // below.
+        let mut cur = this.inner().weak.load(Relaxed);
 
+        loop {
             // check if the weak counter is currently "locked"; if so, spin.
             if cur == usize::MAX {
```
Comment (Connorcpu) [inline src/liballoc/arc.rs]: If `cur` is ever usize::MAX, won't this just keep spinning forever and never reload the value of `cur`?

Comment (Amanieu) [inline src/liballoc/arc.rs]: Ah, good point, I'll make a PR to fix that.

**Pull request merged**